### PR TITLE
Add model tests

### PR DESCRIPTION
**Pull request by aajdinov**
Wrote tests for all models and updated mock data.

### CODE REVIEW

File: lib/fog/profitbricks/models/compute/server.rb [hunk 1]
```diff
@@ -175,7 +176,7 @@ def list_volumes
           volumes = result.body['items'].each { |volume| volume['datacenter_id'] = datacenter_id }
           result.body['items'] = volumes
 
-          result.body
+          result.body['items']
```
Comment (plribeiro3000) [inline lib/fog/profitbricks/models/compute/server.rb]: This is breaking the current behaviour. If we do follow this pattern we need to do a major bump on `fog-profitbricks` so other users that rely on this wont be surprised when they bump the version.

Whats the data returned in this call? Only `items` key?

File: lib/fog/profitbricks/models/compute/snapshot.rb [hunk 2]
```diff
@@ -55,7 +55,7 @@ def update
           options[:discVirtioHotUnplug] = disc_virtio_hot_unplug if disc_virtio_hot_unplug
           options[:discScsiHotPlug]     = disc_scsi_hot_plug if disc_scsi_hot_plug
           options[:discScsiHotUnplug]   = disc_scsi_hot_unplug if disc_scsi_hot_unplug
-          options[:licenceType]         = license_type if license_type
+          options[:licenceType]         = licence_type if licence_type
```
Comment (plribeiro3000) [inline lib/fog/profitbricks/models/compute/snapshot.rb]: What did change here exactly?

File: lib/fog/profitbricks/requests/compute/get_all_lans.rb [hunk 3]
```diff
@@ -93,6 +93,8 @@ def get_all_lans(datacenter_id)
                         :method  => 'GET',
                         :path    => "/datacenters/#{datacenter_id}/lans?depth=1"
                     )
+                rescue => error
+                    Fog::Errors::NotFound.new(error)
```
Comment (plribeiro3000) [inline lib/fog/profitbricks/requests/compute/get_all_lans.rb]: The documentation does not states what happens when no lans are encountered but i guess it does return an empty hash, right?

If so why would we need to rescue exceptions here and thrown a not find error? Wouldn't that be more confusing?

File: lib/fog/profitbricks/requests/compute/get_attached_cdrom.rb [hunk 4]
```diff
@@ -48,6 +48,8 @@ def get_attached_cdrom(datacenter_id, server_id, cdrom_image_id)
               :method  => 'GET',
               :path    => "/datacenters/#{datacenter_id}/servers/#{server_id}/cdroms/#{cdrom_image_id}?depth=1"
           )
+        rescue => error
+          Fog::Errors::NotFound.new(error)
```
Comment (plribeiro3000) [inline lib/fog/profitbricks/requests/compute/get_attached_cdrom.rb]: Same thing here. What does this api return in case of not having any attached CD ROM?

File: tests/profitbricks/models/compute/compute_tests.rb [hunk 5]
```diff
@@ -7,59 +7,538 @@
...
+                                       		:type => 'HDD')
+			volume.wait_for { ready? }
+
+			@volume_id = volume.id
+
+			volume.name 		== 'fog-demo-volume'
+			volume.size 		== 5
+			volume.type			== 'HDD'
+			volume.licence_type	== 'OTHER'
+	    end
```
Comment (plribeiro3000) [inline tests/profitbricks/models/compute/compute_tests.rb]: You might be just following the conventions on the repo but could you use 2 spaces only please?

File: lib/fog/profitbricks/models/compute/server.rb (hunk 1 above)
Comment (aajdinov) [inline lib/fog/profitbricks/models/compute/server.rb]: Thank you for your feedback.

This was reverted in the latest commit. It was an oversight on my end.

File: lib/fog/profitbricks/models/compute/snapshot.rb (hunk 2 above)
Comment (aajdinov) [inline lib/fog/profitbricks/models/compute/snapshot.rb]: The attribute is named 'licence_type' - 'license_type' was a type-o, so I fixed it.

File: lib/fog/profitbricks/requests/compute/get_all_lans.rb (hunk 3 above)
Comment (aajdinov) [inline lib/fog/profitbricks/requests/compute/get_all_lans.rb]: I am aware of the flaws in the current implementation of error handling - it will be revised and updated on a separate branch.

File: tests/profitbricks/models/compute/compute_tests.rb (hunk 5 above)
Comment (aajdinov) [inline tests/profitbricks/models/compute/compute_tests.rb]: It was changed to 2 spaces in the latest commit.

Thank you for your feedback.

File: lib/fog/profitbricks/models/compute/snapshot.rb (hunk 2 above)
Comment (plribeiro3000) [inline lib/fog/profitbricks/models/compute/snapshot.rb]: Damm. so small i didn't notice. lol

Comment (plribeiro3000) [inline lib/fog/profitbricks/models/compute/snapshot.rb]: Good catch

File: lib/fog/profitbricks/requests/compute/get_all_lans.rb (hunk 3 above)
Comment (plribeiro3000) [inline lib/fog/profitbricks/requests/compute/get_all_lans.rb]: Hmm. I'm afraid this breaks the current behaviour. It might be wrong but as of now users are able to rescue each kind of error separately and with this new implementation they wont.
If you are going to change it later anyways, maybe we can leave it as is for now?

File: lib/fog/profitbricks/models/compute/server.rb (hunk 1 above)
Comment (plribeiro3000) [inline lib/fog/profitbricks/models/compute/server.rb]: Gotcha.

File: tests/profitbricks/models/compute/compute_tests.rb [hunk 6]
```diff
@@ -55,6 +57,483 @@
 	  	location.name == 'lasvegas'
     end
 
+    tests('should create a volume').succeeds do
+  		volume = compute.volumes.create(:datacenter_id => @datacenter_id,
+  																		:name => 'fog-demo-volume',
+                                    	:size => 5,
+                                   		:licence_type => 'OTHER',
+                                   		:type => 'HDD')
+			volume.wait_for { ready? }
```
Comment (plribeiro3000) [inline tests/profitbricks/models/compute/compute_tests.rb]: This does not seem to be indented. Maybe you need to check your editor?

File: tests/profitbricks/requests/compute/server_tests.rb [hunk 7]
```diff
@@ -228,20 +285,39 @@
...
+
+    if ENV["FOG_MOCK"] == "false"
+      tests('#get_attached_cdrom').data_matches_schema(@resource_schema) do
         sleep(60)
+        getAttachedVolumeResponse = service.get_attached_cdrom(@datacenter_id, @server_id, @cdrom_id)
+        getAttachedVolumeResponse.body
+      end
+    end
+
+    if ENV["FOG_MOCK"] == "true"
```
Comment (plribeiro3000) [inline tests/profitbricks/requests/compute/server_tests.rb]: You should be able to use [`Fog.mock?`](https://github.com/fog/fog-core/blob/188c1253bd51a9b1aeba7e3ab675de8e677aaada/lib/fog/core/mock.rb#L15) here.

File: tests/profitbricks/requests/compute/nic_tests.rb [hunk 8]
```diff
@@ -44,11 +44,32 @@
...
+        else
+          if image['properties']
+            image['properties']['location'] == 'us/las' &&
+            image['properties']['imageType'] == 'CDROM' &&
+            image['properties']['licenceType'] == 'UNKNOWN'
+          else
+            image['location'] == 'us/las' &&
+            image['imageType'] == 'CDROM' &&
+            image['licenceType'] == 'UNKNOWN'
+          end
```
Comment (plribeiro3000) [inline tests/profitbricks/requests/compute/nic_tests.rb]: Why is the response from the live request and the mocked one different?

File: tests/profitbricks/requests/compute/server_tests.rb [hunk 9]
```diff
@@ -57,22 +85,51 @@
...
+    if ENV["FOG_MOCK"] == "false"
+      tests('#get_image').data_matches_schema(@resource_schema) do
+        getImageResponse = service.get_image(@image_id)
+        getImageResponse.body
+      end
     end
 
     if ENV["FOG_MOCK"] == "true"
-      tests('#update_image').data_matches_schema(@resource_schema) do
+      tests('#update_image').succeeds do
```
Comment (plribeiro3000) [inline tests/profitbricks/requests/compute/server_tests.rb]: Why did you change this test? It now does a different assertion then the original. Was that by purpose?

File: tests/profitbricks/requests/compute/nic_tests.rb (hunk 8 above)
Comment (plribeiro3000) [inline tests/profitbricks/requests/compute/nic_tests.rb]: Why is mock working differently from live requests? They were supposed to work the same way.

File: lib/fog/profitbricks/requests/compute/list_attached_volumes.rb [hunk 10]
```diff
@@ -52,6 +52,8 @@ def list_attached_volumes(datacenter_id, server_id)
               :method  => 'GET',
               :path    => "/datacenters/#{datacenter_id}/servers/#{server_id}/volumes?depth=1"
           )
+        rescue => error
+          Fog::Errors::NotFound.new(error)
```
Comment (plribeiro3000) [inline lib/fog/profitbricks/requests/compute/list_attached_volumes.rb]: You forgot this one here. 😬

File: tests/profitbricks/requests/compute/nic_tests.rb (hunk 8 above)
Comment (aajdinov) [inline tests/profitbricks/requests/compute/nic_tests.rb]: I do not create an image - see [API documentation](https://devops.profitbricks.com/api/cloud/v2/#images).

Therefor, when live testing, I use a pre-exising image from ProfitBricks. I do not want to change any properties on their pre-existing images, which other people might be using.

In case of mock data, I am able to test the update, because it won't will not affect anyone. And because the mock data was updated/changed in one of the previous tests, it will not be the same as the live data on ProfitBricks servers - which was not modified.

File: lib/fog/profitbricks/compute.rb [hunk 11]
```diff
@@ -148,17 +148,22 @@ def request(params)
 
           begin
             response = @connection.request(params)
+            
           rescue Excon::Errors::Unauthorized => error
+            puts 'Unauthorized error'
```
Comment (plribeiro3000) [inline lib/fog/profitbricks/compute.rb]: You can use `Fog` warnings here instead of plain puts.

See [here](https://github.com/fog/fog-core/blob/f6ef6c3d9458bc235be57100acdced264e85e52a/lib/fog/core/logger.rb#L26)

**Comment by plribeiro3000** [comment]
@edevenport We need to be more careful not to merge stuff that are still on review.

I might be wrong but this PR broke several API's. Lets make sure we bump the major version.

Also it is a good idea to ask the person to fix its git config so we can track down who did a change later. Those commits can get lost in the future because of this.

**Comment by edevenport** [comment]
@plribeiro3000 - This PR is based off the ProfitBricks REST API v2, but it mostly focused on models tests which lead to error handling adjustments based on the review. The next PR (closed due to rebase request) is based on their REST API v3. The changes are not drastic, but it will indeed result in a major version bump for the next official release.

I will ask the person about their git config as well as the rebase. I would like to roll out a new release based on the v3 API soon if possible. I'll also get the rubocop issue sorted out for that.

**Comment by plribeiro3000** [comment]
👍
